### PR TITLE
Motion Detection resolves #21 resolves #33 resolves #5 

### LINKED ISSUE #5: As a home security camera user, I want to be able to record footage from my security camera so that I can review it later.

### LINKED ISSUE #21: Motion Detection
Subtask from #2  

Description:
Raspberry Pi should be able to detect when an event happens in the camera's foreground.

### LINKED ISSUE #33: access video feed through browser on network
For ease of use and convenience, users should be able to access the home monitoring system from a web browser on their network.

### PR DESCRIPTION
**Pull request by melippmann**
motiondetect.py is a simple program that reads from a webcam processes image frames and compares them to detect motion, if motion is detected a box is drawn around the moving object and the status of detection is updated. resolves #21

### CODE REVIEW

**Comment by DJ-IRL** [comment]
fix CI

**Review by DJ-IRL — APPROVED** [review]
looks good to me, but we should find a way to zipper this into the Django project asap

File: motiondetect.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    elif len(cnts) == 3:
+        cnts = cnts[1]
+    else:
+        print("something went wrong, exiting program :(")
+        exit()
+    return avg, cnts
+#get background subtraction with k nearest neighbor
+backSubknn = cv.createBackgroundSubtractorKNN()
+#innitial video capture: source ov vid or file path/name
+capture = cv.VideoCapture(0)
```
Comment (kevdouglass) [inline motiondetect.py:43]: create variables that handle static/hardcoded values. For example this 0th index would specify FIRST=0 and pass FIRST in cv.VideoCapture( .. ) function. This way, when we put actual data, it is simple to refactor these values.

File: motiondetect.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    #return resize frame to particular dimension
+    return cv.resize(frame, dimensions, interpolation=cv.INTER_AREA)
+
+#convert frame to grey, compute Gaussian blur for noise reduction, update ave,
+#compute weighted averages, compute difference between wieghted ave and grayscale frame to get delta (background bodel - grayscale frame)
+def imgProcess(frame, avg):
+    #convert to grayscale image
+    gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
+    #blur image to reduce noise (filter size 21X21)
+    gray = cv.GaussianBlur(gray, (11,11),0)
```
Comment (kevdouglass) [inline motiondetect.py:19]: create variables to handle harcoded values so they are easier to change, especially for readabliity purposes

File: motiondetect.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    #if first loop, need to set avg
+    if avg is None:
+        avg = gray.copy().astype("float")
+    #get weighted average of prev frame
+    cv.accumulateWeighted(gray, avg, 0.5)
```
Comment (kevdouglass) [inline motiondetect.py:24]: variable name rather than value 0.5 for readability

File: motiondetect.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    #compute difference between first frame and cur frame
+    frameDelta = cv.absdiff(gray, cv.convertScaleAbs(avg))
+    thresh = cv.threshold(frameDelta, 5, 255, cv.THRESH_BINARY)[1]
+    #dilate the threshold image to fill in holes
+    thresh = cv.dilate(thresh, None, iterations=2)
+    #get contours
+    cnts = cv.findContours(thresh.copy(), cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
+    if len(cnts) == 2:
+        cnts = cnts[0]
```
Comment (kevdouglass) [inline motiondetect.py:33]: swap hardcoded values for variable names

File: motiondetect.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    elif len(cnts) == 3:
```
Comment (kevdouglass) [inline motiondetect.py:34]: swap hardcoded values for variable names

File: motiondetect.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        #set motion box on frame
+        cv.rectangle(frame_resized, (x,y), (x+w, y+h), (0, 0, 255), 2)
+        #change notification text
+        text = "Motion!"
+    #add text to frame
+    cv.putText(frame_resized, "Status: {}".format(text), (15,15), cv.FONT_HERSHEY_SIMPLEX, .5, (0,0,255), 1)
+    #display the image
+    cv.imshow('Video', frame_resized)
+    #check for interupt
+    if cv.waitKey(28) & 0xFF==ord('d'):
```
Comment (kevdouglass) [inline motiondetect.py:70]: swap hardcoded values for variable names. especially if the same values are used in multiple places (RGB values), (0.5), (1),(15,15)..

File: motiondetect.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+capture = cv.VideoCapture(0)
+#initially the average frame is 0, set in loop from grayscale image
+avg = None
+#infinate loop and capture video until 'd' is pressed
+while True:
+    text = "searching..."
+    #read video, gets a bool and frame
+    isTrue, frame = capture.read()
+    #resize frame
+    frame_resized =rescaleFrame(frame, 1)
```
Comment (kevdouglass) [inline motiondetect.py:52]: swap hardcoded values for variable names

File: motiondetect.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    frameDelta = cv.absdiff(gray, cv.convertScaleAbs(avg))
+    thresh = cv.threshold(frameDelta, 5, 255, cv.THRESH_BINARY)[1]
+    #dilate the threshold image to fill in holes
+    thresh = cv.dilate(thresh, None, iterations=2)
+    #get contours
+    cnts = cv.findContours(thresh.copy(), cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
+    if len(cnts) == 2:
+        cnts = cnts[0]
+    elif len(cnts) == 3:
+        cnts = cnts[1]
```
Comment (kevdouglass) [inline motiondetect.py:35]: Please change hardcoded integer/double/float values for variable names for reusablitiy.

File: motiondetect.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    #blur image to reduce noise (filter size 21X21)
+    gray = cv.GaussianBlur(gray, (11,11),0)
+    #if first loop, need to set avg
+    if avg is None:
+        avg = gray.copy().astype("float")
+    #get weighted average of prev frame
+    cv.accumulateWeighted(gray, avg, 0.5)
+    #compute difference between first frame and cur frame
+    frameDelta = cv.absdiff(gray, cv.convertScaleAbs(avg))
+    thresh = cv.threshold(frameDelta, 5, 255, cv.THRESH_BINARY)[1]
```
Comment (kevdouglass) [inline motiondetect.py:27]: swap hardcoded values for variable names

**Review by ConnorGanaway — APPROVED** [review]

**Pull request merged**